### PR TITLE
Restart scanner after cancel

### PR DESCRIPTION
**Pull request by kain88-de**
https://bugs.launchpad.net/mixxx/+bug/1425705

This shifts the responsibility to notify about hashed AND scanned directories to the importFiles task. I think this is the clearest way to do this. I personally don't care much about the extra work of scheduling another task.

@rryan do you think there is a better way to do this?

### CODE REVIEW

**Comment by ywwg** [comment]
What testing have you done to check that this works?  The code is clean enough but I don't really understand the issues involved

File: src/library/scanner/libraryscanner.cpp [hunk 1]
```diff
@@ -402,10 +402,10 @@ void LibraryScanner::queueTask(ScannerTask* pTask) {
     m_pool.start(pTask);
 }
 
-void LibraryScanner::directoryHashed(const QString& directoryPath,
-                                     bool newDirectory, int hash) {
-    ScopedTimer timer("LibraryScanner::directoryHashed");
-    // qDebug() << "LibraryScanner::directoryHashed" << directoryPath
+void LibraryScanner::directoryHashedAndScanned(const QString& directoryPath,
```
Comment (kain88-de) [inline src/library/scanner/libraryscanner.cpp]: This function is the issue. It assumed that a directory must have been scanned as well once it was hashed.

File: src/library/scanner/recursivescandirectorytask.cpp [hunk 2]
```diff
@@ -79,14 +79,11 @@ void RecursiveScanDirectoryTask::run() {
...
-                                                      filesToImport, possibleCovers,
-                                                      m_pToken));
-        }
+        m_pScanner->queueTask(new ImportFilesTask(m_pScanner, m_scannerGlobal,
+                                                  dirPath, newHash, prevHashExists,
+                                                  filesToImport, possibleCovers,
+                                                  m_pToken));
 
-        // Insert or update the hash in the database.
-        emit(directoryHashed(dirPath, !prevHashExists, newHash));
```
Comment (kain88-de) [inline src/library/scanner/recursivescandirectorytask.cpp]: But what actually happens after a directory is hashed is that the files are scheduled for importing in the task queue. If a scan is canceled all pending task are aborted.  This means mixxx thought the files were imported even if the task was canceled

**Comment by kain88-de** [comment]
You can check the bug by canceling a library scan and looking at the sql. Some entries in the library hash table should be marked as need verification.  In the current master no directory is marked at needs verification but with this patch every directory where the files were not imported are marked as need verification. I hope that description is clear.

**Comment by ywwg** [comment]
we already have a few library tests, is there any way we can add one for this?  Otherwise this seems low-risk enough to merge

File: src/library/scanner/recursivescandirectorytask.cpp [hunk 3]
```diff
@@ -79,14 +79,11 @@ void RecursiveScanDirectoryTask::run() {
     if (prevHash != newHash) {
         // Rescan that mofo! If importing fails then the scan was cancelled so
         // we return immediately.
-        if (!filesToImport.isEmpty()) {
```
Comment (rryan) [inline src/library/scanner/recursivescandirectorytask.cpp]: Instead, could you keep the ```emit(directoryHashed(...))``` for the empty case? That would prevent scheduling tons of no-op tasks if the user has a bunch of directories with no supported files for some reason.

Comment (kain88-de) [inline src/library/scanner/recursivescandirectorytask.cpp]: I thought about that. I lke that in my solution only one class had the responsibility to emit a signal. I think that makes the code easier to understand. .(the whole task scheduling here is already hard to follow). 

I haven't noticed that the no-op task would slow down the scanning for me. So I would like to keep it that way unless it is a huge performance hit for older/slow hardware.

**Comment by kain88-de** [comment]
I added a signal when no files have to be importet.

@ywwg 

I would like to but I have no idea how and even if we can test the interacting parts of the scanner queueing system.

**Comment by ywwg** [comment]
Since this is a large bug, I'm definitely leaning more toward merging this than waiting for perfect tests.  For the no-op case, it looks like you're emitted directoryHashed, but nothing is connected to that signal any more.  Is that right?

**Comment by kain88-de** [comment]
'directoryHashed' connects to 'directoryHashedAndScanned' in the scannertask constructor.

**Comment by kain88-de** [comment]
@ywwg 

If you know how we can test the mutliple threads and their interconnection I'm all ears. But currently I'm not aware how this can be done in unit-test. The only way I know is to use the actual application.

**Comment by ywwg** [comment]
Yeah I agree I don't think this is testable right now.  But it looks like you addressed all of RJ's concerns so I'm willing to say LGTM
